strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java

```diff
@@ -39,29 +47,99 @@ public void postConstruct()
-    protected void cascadeEntityIdentification(T entity)
+    protected <S extends T> S cascadeEntitySave(T entity)
     {
         identifyEntity(entity);
-        ReflectionUtils.doWithFields(entity.getClass(), (field)->{
+        
+        ReflectionUtils.doWithFields(entity.getClass(), (field) -> {
             ReflectionUtils.makeAccessible(field);
-            Class<? extends GenericEntity> t = (Class<? extends GenericEntity>) field.getType();
-            if (!GenericEntity.class.isAssignableFrom(t))
+
+            Set<CascadeType> cascadeTypeSet = Arrays.stream(field.getAnnotations())
+                                                    .map(a -> exposeCascadeType(a))
+                                                    .reduce((c1,
+                                                             c2) -> {
+                                                        c1.addAll(c2);
+                                                        return c1;
+                                                    })
+                                                    .orElse(Collections.emptySet());
+            
+            if (!cascadeTypeSet.stream().anyMatch(c -> CascadeType.ALL.equals(c) || CascadeType.MERGE.equals(c)
+                    || CascadeType.PERSIST.equals(c)))
             {
                 return;
             }
             
-            GenericEntity subEntity = (GenericEntity) ReflectionUtils.getField(field, entity);
-            if (subEntity == null)
+            Class<?> fieldType = field.getType();
+            Object fieldValue = ReflectionUtils.getField(field, entity);
+
+            if (fieldValue == null)
             {
                 return;
             }
 
-            CommonCrudService<GenericEntity> entityService = (CommonCrudService<GenericEntity>) entityServiceRegistry.getEntityService(subEntity.getClass());
-            entityService.identifyEntity(subEntity);
+            if (Collection.class.isAssignableFrom(fieldType))
+            {
+                Collection<Object> collection = (Collection<Object>) fieldValue;
+                List<Object> replaceCollection = new LinkedList<>();
+                collection.removeIf(a -> {
+                    Object b = tryToCascadeEntitySave(a);
+                    if (b != a)
+                    {
+                        replaceCollection.add(b);
+                        return true;
+                    }
+                    return false;
+                });
+                collection.addAll(replaceCollection);
+            }
+            else
+            {
+                Object newFieldValue = tryToCascadeEntitySave(fieldValue);
+                if (newFieldValue != fieldValue)
+                {
+                    ReflectionUtils.setField(field, entity, newFieldValue);
+                }
+            }
+
         });
+
+        return getDelegate().save(entity);
     }
 
+    private Set<CascadeType> exposeCascadeType(Annotation a)
```

Comment:
Consider new class for this utility method

Reply:
yeah, it can be utility method in the future